core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -443,9 +466,27 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorIdle(executorId: String): Unit = synchronized {
     if (executorIds.contains(executorId)) {
       if (!removeTimes.contains(executorId) && !executorsPendingToRemove.contains(executorId)) {
+
+        val hasCachedBlocks =
+          executorsWithCachedBlocks.contains(executorId) ||
+            executorEndpoints.get(executorId).exists(_.askWithRetry[Boolean](HasCachedBlocks))
+
+        if (hasCachedBlocks) executorsWithCachedBlocks += executorId
```

Comment:
Not sure what you mean. This update is happens when the executor becomes idle (immediately after the last scheduled task is done). We don't actually remove it until the timeout anyway. 

The map can be removed, but that would mean one RPC call each time an executor becomes idle - even if we know that the executor has cached blocks. We don't actually remove from the map exactly for that reason - since we know the executor has cached blocks and we should not get rid of it. The size is pretty bounded anyway - the max possible size = max executors.